sklearn/utils/estimator_checks.py

```diff
@@ -602,25 +786,81 @@ def check_estimator(estimator=None, generate_only=False, *, legacy: bool = True)
 
     name = type(estimator).__name__
 
-    def checks_generator():
-        # we first need to check if the estimator is cloneable for the rest of the tests
-        # to run
-        yield estimator, partial(check_estimator_cloneable, name)
-        for check in _yield_all_checks(estimator, legacy=legacy):
-            for check_instance in _yield_instances_for_check(check, estimator):
-                maybe_skipped_check = _maybe_skip(check_instance, check)
-                yield check_instance, partial(maybe_skipped_check, name)
-
+    # TODO(1.8): remove generate_only
     if generate_only:
-        return checks_generator()
+        warnings.warn(
+            "`generate_only` is deprecated in 1.6 and will be removed in 1.8. "
+            "Use :func:`~sklearn.utils.estimator_checks.estimator_checks` instead.",
+            FutureWarning,
+        )
+        return estimator_checks_generator(
+            estimator, legacy=legacy, expected_failed_checks=None, mark="skip"
+        )
 
-    for estimator, check in checks_generator():
+    test_results = []
+    failed = False
+
+    for estimator, check in estimator_checks_generator(
+        estimator,
+        legacy=legacy,
+        expected_failed_checks=expected_failed_checks,
+        # Not marking tests to be skipped here, we run and simulate an xfail behavior
+        mark=None,
+    ):
+        test_can_fail, reason = _should_be_skipped_or_marked(
+            estimator, check, expected_failed_checks
+        )
         try:
             check(estimator)
```

Comment:
Shouldn't we also record checks that passed successfully in `test_results`?

Reply:
And since we pass the "status" as an argument to the callback, we could as well, `on_result` generic callable that would always be called: for passed, failed, xfailed or skipped checks instead of just calling the callback for "failed" and "xfailed" cases.

This would add maximum flexibility and homogeneity with minimum extra code.